### PR TITLE
PIM-6350: As Julia, I would like to change the parent of a variant product/sub product model from the UI

### PR DESCRIPTION
**Pull request by rouliane**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | OK
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | OK
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: tests/legacy/features/Pim/Behat/Context/MassEditJobContext.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+use Behat\Gherkin\Node\TableNode;
+use Context\Spin\SpinCapableTrait;
+
+final class MassEditJobContext extends PimContext implements SnippetAcceptingContext
+{
+    use SpinCapableTrait;
+
+    private const USERNAME_FOR_JOB_LAUNCH = 'admin';
+
+    private const MASS_CHANGE_PARENT_JOB_NAME = 'change_parent_product';
```
Comment (rouliane) [inline tests/legacy/features/Pim/Behat/Context/MassEditJobContext.php]: Specific job name but generic class name.

TODO : improve this part...

**Comment by rouliane** [comment]
TODO : add changelog

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace Pim\Bundle\EnrichBundle\Connector\Processor\MassEdit\Product;
+
+use Akeneo\Component\StorageUtils\Updater\ObjectUpdaterInterface;
+use Pim\Bundle\EnrichBundle\Connector\Processor\AbstractProcessor;
+use Pim\Component\Catalog\Model\EntityWithFamilyVariantInterface;
+use Pim\Component\Catalog\Model\ProductModelInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+class ChangeParentProcessor extends AbstractProcessor
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php]: phpdoc please

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        ObjectUpdaterInterface $productUpdater,
+        ObjectUpdaterInterface $productModelUpdater
+    ) {
+        $this->productValidator = $productValidator;
+        $this->productModelValidator = $productModelValidator;
+        $this->productUpdater = $productUpdater;
+        $this->productModelUpdater = $productModelUpdater;
+    }
+
+    public function process($product)
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php]: : ?EntityWithFamilyVariantInterface

Please typehint all you methods :-)

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php (hunk 2 above)
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php]: Add a comment to describe your class and licence etc.

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php [hunk 4]
```diff
@@ -391,6 +395,28 @@ public function removeAction(Request $request, $id): Response
...
+    {
+        $familyVariant = $this->familyVariantRepository->findOneByIdentifier($familyVariantCode);
+        if (null === $familyVariant) {
+            throw new \InvalidArgumentException(sprintf('Unknown family variant code "%s"', $familyVariantCode));
+        }
+
+        return $familyVariant;
+    }
+
+    private function buildNormalizedProductModeldsList(array $productModels): array
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php]: `normalizePorductModels` seems enough for me.

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    /** @var ValidatorInterface */
+    protected $productModelValidator;
+
+    /** @var ObjectUpdaterInterface */
+    protected $productUpdater;
+
+    /** @var ObjectUpdaterInterface */
+    protected $productModelUpdater;
+
+    public function __construct(
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php]: You forgot comment on all methods of this class ;)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/change-parent-product.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
+'use strict';
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/change-parent-product.js]: This is a copy of `add-to-existing-product-model.js`. Don't do that :-). Just rename the original one so the name fit the usage (something like `update-entity-with-family-variant-parent.js`), and create your own module with its specific configuration.

File: src/Pim/Bundle/EnrichBundle/spec/Connector/Processor/MassEdit/Product/ChangeParentProcessorSpec.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
+<?php
+
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/spec/Connector/Processor/MassEdit/Product/ChangeParentProcessorSpec.php]: declare(strict_type=1);

File: src/Pim/Component/Catalog/Repository/ProductModelRepositoryInterface.php [hunk 8]
```diff
@@ -94,6 +94,13 @@ public function findSubProductModels(FamilyVariantInterface $familyVariant): arr
      */
     public function findRootProductModels(FamilyVariantInterface $familyVariant): array;
 
+    /**
+     * @param FamilyVariantInterface $familyVariant
+     *
+     * @return array
+     */
+    public function findProductModelsForFamilyVariant(FamilyVariantInterface $familyVariant): array;
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Repository/ProductModelRepositoryInterface.php]: To add to the BC breacks.

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            return null;
+        }
+
+        return $product;
+    }
+
+    private function validateProduct($product)
+    {
+        if (! $product instanceof EntityWithFamilyVariantInterface) {
+            throw new \InvalidArgumentException('The product is not correct');
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/MassEdit/Product/ChangeParentProcessor.php]: Add a better message pls. Like "Given entity shoudl be an instance of EntityWithToto. Instance of 'anotherwrongclass'  given."

File: tests/legacy/features/Pim/Behat/Context/MassEditJobContext.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
+<?php
+
```
Comment (damien-carcel) [inline tests/legacy/features/Pim/Behat/Context/MassEditJobContext.php]: declare(strict_type=1);
And typehing the methods

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php [hunk 11]
```diff
@@ -359,12 +357,18 @@ public function searchLastLevelProductModelByCode(Request $request): JsonRespons
...
-            $normalizedProductModels[$productModel->getCode()] = $this->normalizeProductModel(
-                $productModel
-            );
-        }
+        $normalizedProductModels = $this->buildNormalizedProductModeldsList($productModels);
+
+        return new JsonResponse($normalizedProductModels);
+    }
+
+    public function listFamilyVariantProductModels(Request $request)
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php]: Php doc

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php [hunk 12]
```diff
@@ -391,6 +395,28 @@ public function removeAction(Request $request, $id): Response
         return new JsonResponse();
     }
 
+    private function getFamilyVariant(string $familyVariantCode): FamilyVariantInterface
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php]: phpdoc

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductModelController.php]: buildNormalizedProductModelDsList => you have a D before sList

File: src/Pim/Bundle/EnrichBundle/Resources/config/connector/processors.yml [hunk 13]
```diff
@@ -94,3 +94,11 @@ services:
             - '@pim_user.provider.user'
             - '@security.token_storage'
             - '@pim_connector.processor.bulk_media_fetcher'
+
+    pim_enrich.connector.processor.mass_edit.product.change.parent:
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Resources/config/connector/processors.yml]: `change_parent` should be better

File: src/Pim/Bundle/EnrichBundle/Resources/config/form_extensions/mass_edit/change_parent_product.yml [hunk 14]
```diff
@@ -0,0 +1,36 @@
+extensions:
+    pim-mass-product-edit-configure-change_parent_product_model-family:
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Resources/config/form_extensions/mass_edit/change_parent_product.yml]: you used - and _, can you use only - pls ? Same for all in this file.

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/change-parent-product.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * Change the parent product model
+ *
+ * @author    Julian Prud'homme <julian.prudhomme@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/change-parent-product.js]: 2018 :)

File: tests/legacy/features/Pim/Behat/Context/MassEditJobContext.php [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    private const USERNAME_FOR_JOB_LAUNCH = 'admin';
+
+    private const MASS_CHANGE_PARENT_JOB_NAME = 'change_parent_product';
+
+    private $productRepository;
+
+    private $productModelRepository;
+
+    public function __construct(string $mainContextClass, IdentifiableObjectRepositoryInterface $productRepository, IdentifiableObjectRepositoryInterface $productModelRepository)
```
Comment (willy-ahva) [inline tests/legacy/features/Pim/Behat/Context/MassEditJobContext.php]: php doc

**Pull request merged**